### ISSUE TITLE
Entry points for 3rd party modules

### ISSUE THREAD

**Issue by robinandeer** [ewels/MultiQC#10]
Someone mentioned what the best way would be to share 3rd party modules at DNA club

Maybe people are getting sick of me recommending **setuptools entry points** but I think they could prove a good fit in this case

Basically it would allow other people to build modules as Python packages that can be installed using pip. Whatever packages are active in the current env are the once used by MultiQC :smiley:

**Comment by robinandeer** [comment]
Fantastic job btw on the whole project - very impressive!

**Comment by marcelm** [comment]
The Pylons documentation has [a section on setuptools entry points](http://pylons-webframework.readthedocs.org/en/latest/advanced_pylons/entry_points_and_plugins.html), which I found very helpful. Seems to fit your use case perfectly.

**Comment by ewels** [comment]
Notes from Robin's help:

Refactored code for `scripts/multiqc`:
```python
entry_points = pkg_resources.iter_entry_points('multiqc.modules.v1')
modules_to_run = [entry_point for entry_point in entry_points if entry_point in run_modules]

for module_ep in modules_to_run:
    try:
        # this is like import
        MultiqcModule = module_ep.load()
        modules_output.append(MultiqcModule(report))
    except UserWarning:
        pass # No samples found
    except:
        logging.error("Error - module {} raised an exception: {}"
                      .format(m, traceback.format_exc()))
```

Code for `setup.py`:

```python
entry_points={
    'multiqc.modules.v1': [
        # the class can now be named whatever ^^
        'bowtie = multiqc.bowtie:MultiqcModule',
        'fastqc = multiqc.fastqc:MultiqcModule'
    ]
}
```

Code for other people's `setup.py`:

```python
entry_points={
    'multiqc.modules.v1': [
        'mypckge = multiqc_mypckge:MypckgeModule'
    ]
}
```

People can then install both by running `pip install multiqc multiqc-mypckge`

Whilst this stuff is undeniably cool, for now I think I'm going to leave it - if nothing else, in an effort to encourage people to pull request their code back to the main repo for everyone else.

I'll leave this issue here for future reference though - feel free to upvote / comment if you would like to see it implemented.

**Comment by ewels** [comment]
Thanks again for this @robinandeer - it's [super cool](https://github.com/ewels/MultiQC/blob/master/setup.py#L47-L65)!

Next up - I'm going to try writing an NGI package to plug into this framework, with a NGI specific template. I love that we'll be able to keep the code entirely separate like this and not mess up the core package...

**Comment by ewels** [comment]
@robinandeer - one thing I couldn't figure out was how to specify the class name in the `setup.py` file as you did. All the examples I could find online were for `def`s rather than classes, don't know if that makes a difference? I'm probably just being slow..

**Comment by ewels** [comment]
Figured it out - I was being dense - should use a `:` for function names, I was trying to use a `.` and it couldn't find the module. Obviously.

**Comment by ewels** [comment]
Fixed in e1dd829a2f5f14a715c646d1fd3556d140a9f93a
